File: reference/sigv4.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""SigV4 签名器（火山引擎兼容）—— **零第三方依赖**。

为什么手写而不是装 botocore
---------------------------
火山引擎 ImageX 的 OpenAPI 用的是 AWS SigV4 **兼容格式**，但签名 scope 是
``region=i18n`` / ``service=imagex``（口径来自真实抓包的 Authorization 头，
见 API.md §5.10 ②）。目标执行节点是一台干净的海外小机器，
不适合为了一个确定性算法拖进 boto3 + botocore 约 60MB 的依赖树。

正确性如何保证
--------------
算法是确定性的，所以可以用 **AWS 官方文档的已知测试向量**离线自证，
不需要任何真实凭据、不产生任何网络请求：

    python sigv4.py

期望输出 `[OK] 5/5 用例通过`。这条自证在拿到新鲜 cookie 之前就能跑，
把「签名对不对」从「上服务器碰运气」变成「本机秒验」。

调用形态
--------
    from sigv4 import sign

    h = sign(
        method="GET",
        url="https://ads.tiktok.com/creative/creativestudio/upload-proxy"
            "?Action=ApplyImageUpload&Version=2018-08-01"
            "&ServiceId=n2703mo9gi&FileSize=45431&s=m0kza0mqao&device_platform=web",
        headers={"host": "ads.tiktok.com"},
        payload=b"",
        access_key=sts["AccessKeyId"],
        secret_key=sts["SecretAccessKey"],
        session_token=sts["SessionToken"],
        when=sts_dt,                 # 用 STS 的 CurrentTime，见下
    )
    req_headers.update(h)
"""

from __future__ import annotations

import datetime as _dt
import hashlib
import hmac
import sys
from typing import Dict, Mapping, Optional, Tuple
from urllib.parse import parse_qsl, quote, urlparse

ALGORITHM = "AWS4-HMAC-SHA256"
TERMINATOR = "aws4_request"

# RFC 3986 unreserved：A-Z a-z 0-9 - _ . ~ 之外一律百分号编码
_UNRESERVED = "-_.~"

# 空载荷的 sha256 —— 每个请求都要带，GET 也用这个
EMPTY_SHA256 = hashlib.sha256(b"").hexdigest()


def uri_encode(value: str, *, encode_slash: bool = True) -> str:
    """RFC 3986 编码，十六进制大写（SigV4 要求大写）。"""
    safe = _UNRESERVED if encode_slash else _UNRESERVED + "/"
    return quote(str(value), safe=safe, encoding="utf-8")


def sha256_hex(data) -> str:
    if isinstance(data, str):
        data = data.encode("utf-8")
    return hashlib.sha256(data).hexdigest()


def canonical_query_string(url: str) -> str:
    """查询串归一化：先编码，再按**编码后的键**做字节序升序。

    ⚠️ 排序必须在编码之后进行 —— 大写字母（`A`=65）排在小写（`s`=115）之前，
    所以 `Action, FileSize, ServiceId, Version, device_platform, s`
    才是正确顺序。按原始串排序会得到不同结果、签名必然错。
    """
    query = urlparse(url).query
    if not query:
        return ""
    pairs = parse_qsl(query, keep_blank_values=True)
    encoded = sorted((uri_encode(k), uri_encode(v)) for k, v in pairs)
    return "&".join("%s=%s" % (k, v) for k, v in encoded)


def _normalize_headers(headers: Mapping[str, str]) -> Dict[str, str]:
    """键小写、值折叠连续空白并去首尾（SigV4 对值的规范化要求）。"""
    out: Dict[str, str] = {}
    for key, value in headers.items():
        out[str(key).lower().strip()] = " ".join(str(value).split())
    return out


def build_canonical_request(method: str, url: str, headers: Mapping[str, str],
                            payload_sha: str) -> Tuple[str, str]:
    """返回 (规范请求串, 已签名头列表)。

    拼接后的形状（注意查询串为空时那一行是空的，且头块末尾会多出一个空行）：

        GET
        /

        content-type:...
        host:iam.amazonaws.com
        x-amz-date:20150830T123600Z

        content-type;host;x-amz-date
        e3b0c442...
    """
    parsed = urlparse(url)
    body: Dict[str, str] = dict(headers)
    if "host" not in {k.lower() for k in body}:
        body["host"] = parsed.netloc

    lower = _normalize_headers(body)
    signed_headers = ";".join(sorted(lower))
    canon_headers = "".join("%s:%s\n" % (k, lower[k]) for k in sorted(lower))

    canonical = "\n".join([
        method.upper(),
        uri_encode(parsed.path or "/", encode_slash=False),
        canonical_query_string(url),
        canon_headers,
        signed_headers,
        payload_sha,
    ])
    return canonical, signed_headers


def derive_signing_key(secret_key: str, date_stamp: str,
                       region: str, service: str) -> bytes:
    """AWS4 + secret → HMAC 链 → 派生签名密钥（4 轮）。"""
    key = ("AWS4" + secret_key).encode("utf-8")
    for msg in (date_stamp, region, service, TERMINATOR):
        key = hmac.new(key, msg.encode("utf-8"), hashlib.sha256).digest()
    return key


def sign(
    *,
    method: str,
    url: str,
    access_key: str,
    secret_key: str,
    headers: Optional[Mapping[str, str]] = None,
    payload=b"",
    session_token: Optional[str] = None,
    region: str = "i18n",
    service: str = "imagex",
    when: Optional[_dt.datetime] = None,
    content_sha256: Optional[str] = None,
    sign_content_sha256: bool = False,
) -> Dict[str, str]:
    """对一次请求签名，返回**需要补进请求头**的字段。

    返回的键一律用标准 HTTP 大小写形式（`Authorization` / `X-Amz-Date` / …），
    可直接 `req.headers.update(...)`。

    `when` 必须是 **UTC**。默认取当前 UTC 时间；但对接 TikTok 时应当显式传入
    STS 的 `CurrentTime`（理由见模块末尾）。

    `sign_content_sha256` —— 是否附带并签名 `X-Amz-Content-Sha256`。
    ⚠️ AWS 规范规定**所有 `x-amz-*` 头都必须进签名**，所以「发不发」和
    「签不签」是同一个开关，不存在「发了但不签」这个中间态。
    - `False`（默认）：完全不发这个头。这是 SigV4 的最小正确形态，
      也是 AWS 官方示例的形态 —— 非 S3 服务通常不需要它。
    - `True`：附带并签名。TikTok 的真实抓包里这个头**是存在的**
      （§5.10 ④ 的 `X-Amz-Content-Sha256`），所以对接 TikTok 时传 `True`。
    `session_token` 一旦提供就必然附带并签名 —— STS 凭据缺它必被拒。
    """
    when = when or _dt.datetime.now(_dt.timezone.utc)
    if when.tzinfo is None:
        when = when.replace(tzinfo=_dt.timezone.utc)
    when = when.astimezone(_dt.timezone.utc)

    amz_date = when.strftime("%Y%m%dT%H%M%SZ")
    date_stamp = when.strftime("%Y%m%d")

    payload_sha = content_sha256 or sha256_hex(payload)

    hdrs: Dict[str, str] = {str(k): str(v) for k, v in (headers or {}).items()}
    for key in list(hdrs):
        if key.lower() in ("authorization", "x-amz-date", "x-amz-security-token",
                           "x-amz-content-sha256"):
            del hdrs[key]
    hdrs.setdefault("host", urlparse(url).netloc)
    hdrs["X-Amz-Date"] = amz_date
    if sign_content_sha256:
        hdrs["X-Amz-Content-Sha256"] = payload_sha
    if session_token:
        hdrs["X-Amz-Security-Token"] = session_token

    canonical, signed_headers = build_canonical_request(method, url, hdrs, payload_sha)

    scope = "%s/%s/%s/%s" % (date_stamp, region, service, TERMINATOR)
    string_to_sign = "\n".join([
        ALGORITHM,
        amz_date,
        scope,
        sha256_hex(canonical),
    ])

    signing_key = derive_signing_key(secret_key, date_stamp, region, service)
    signature = hmac.new(signing_key, string_to_sign.encode("utf-8"),
                         hashlib.sha256).hexdigest()

    out = {
        "Authorization": (
            "%s Credential=%s/%s, SignedHeaders=%s, Signature=%s"
            % (ALGORITHM, access_key, scope, signed_headers, signature)
        ),
        "X-Amz-Date": amz_date,
    }
    if sign_content_sha256:
        out["X-Amz-Content-Sha256"] = payload_sha
    if session_token:
        out["X-Amz-Security-Token"] = session_token
    return out


# --------------------------------------------------------------------------
# 自证：AWS 官方文档的已知测试向量
# --------------------------------------------------------------------------
# 来源：AWS 文档 "Examples of the complete Version 4 signing process" 里的
# ListUsers 示例。凭据与签名都是公开的示例值，不是真实密钥。
#
# 之所以能拿它当脚手架：火山引擎是 SigV4 *兼容* 实现，签名算法本身完全相同，
# 只有 scope 里的 region/service 两个字符串不同。所以只要这组向量过了，
# 就能确定 canonical request / 密钥派生 / 签名三处都没有写错，
# 换到 i18n/imagex 只是代入不同的 scope 字符串。
# --------------------------------------------------------------------------

_AWS_KEY_ID = "AKIDEXAMPLE"
_AWS_SECRET = "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY"
_AWS_WHEN = _dt.datetime(2015, 8, 30, 12, 36, 0, tzinfo=_dt.timezone.utc)


def _case_listusers() -> Tuple[bool, str, str]:
    """官方 ListUsers 示例 —— authoritative。"""
    url = "https://iam.amazonaws.com/?Action=ListUsers&Version=2010-05-08"
    headers = {
        "content-type": "application/x-www-form-urlencoded; charset=utf-8",
        "host": "iam.amazonaws.com",
    }
    out = sign(method="GET", url=url, headers=headers, payload=b"",
               access_key=_AWS_KEY_ID, secret_key=_AWS_SECRET,
               region="us-east-1", service="iam", when=_AWS_WHEN)
    expect = ("AWS4-HMAC-SHA256 Credential=AKIDEXAMPLE/20150830/us-east-1/iam/aws4_request, "
              "SignedHeaders=content-type;host;x-amz-date, "
              "Signature=5d672d79c15b13162d9279b0855cfba6789a8edb4c82c400e06b5924a6f2b5d7")
    return out["Authorization"] == expect, out["Authorization"], expect


def _case_empty_payload_hash() -> Tuple[bool, str, str]:
    """GET 空载荷的 sha256 必须等于 e3b0c442…（写错成 '' 是最常见的坑）。"""
    expect = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
    return EMPTY_SHA256 == expect, EMPTY_SHA256, expect


def _case_query_sorting() -> Tuple[bool, str, str]:
    """编码后按字节序排序：大写键排在小写键之前。"""
    url = ("https://x/?s=zzz&device_platform=web&Version=2018-08-01"
           "&ServiceId=n2703mo9gi&FileSize=45431&Action=ApplyImageUpload")
    got = canonical_query_string(url)
    expect = ("Action=ApplyImageUpload&FileSize=45431&ServiceId=n2703mo9gi"
              "&Version=2018-08-01&device_platform=web&s=zzz")
    return got == expect, got, expect


def _case_tiktok_url_shape() -> Tuple[bool, str, str]:
    """真实 TikTok 代理 URL 的规范化结果（含空值参数保留）。"""
    url = ("https://ads.tiktok.com/creative/creativestudio/upload-proxy"
           "?Action=ApplyImageUpload&Version=2018-08-01"
           "&ServiceId=n2703mo9gi&FileSize=45431&s=m0kza0mqao&device_platform=web")
    got = canonical_query_string(url)
    expect = ("Action=ApplyImageUpload&FileSize=45431&ServiceId=n2703mo9gi"
              "&Version=2018-08-01&device_platform=web&s=m0kza0mqao")
    return got == expect, got, expect


def _case_signing_key_vector() -> Tuple[bool, str, str]:
    """密钥派生链的已知输出（同来源于 AWS 文档示例）。"""
    key = derive_signing_key("wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY",
                             "20120215", "us-east-1", "iam")
    expect = "f4780e2d9f65fa895f9c67b32ce1baf0b0d8a43505a000a1a9e090d414db404d"
    return key.hex() == expect, key.hex(), expect


def _case_content_sha256_override() -> Tuple[bool, str, str]:
    """显式传入 content_sha256 时必须以它为准，不能被重算覆盖。"""
    url = "https://ads.tiktok.com/creative/creativestudio/upload-proxy?Action=CommitImageUpload"
    payload = b'{"SessionKey":"abc"}'
    forced = sha256_hex(payload)
    out = sign(method="POST", url=url, payload=payload, content_sha256=forced,
               sign_content_sha256=True,
               access_key=_AWS_KEY_ID, secret_key=_AWS_SECRET, when=_AWS_WHEN)
    ok = out["X-Amz-Content-Sha256"] == forced and forced != EMPTY_SHA256
    return ok, out["X-Amz-Content-Sha256"], forced


def _case_amz_headers_must_be_signed() -> Tuple[bool, str, str]:
    """AWS 铁律：凡出现在请求里的 `x-amz-*` 头，必须出现在 SignedHeaders 里。

    这是「发了但不签」这个中间态不存在的证据 —— 也正是本次自证抓出来的缺陷：
    原实现无条件附带 `X-Amz-Content-Sha256` 并签名，会让 AWS 官方 IAM 向量的
    签名对不上（那个示例根本不发这个头）。
    """
    url = "https://ads.tiktok.com/creative/creativestudio/upload-proxy?Action=ApplyImageUpload"
    with_both = sign(method="GET", url=url, payload=b"",
                     sign_content_sha256=True, session_token="STS2example",
                     access_key=_AWS_KEY_ID, secret_key=_AWS_SECRET, when=_AWS_WHEN)
    minimal = sign(method="GET", url=url, payload=b"",
                   access_key=_AWS_KEY_ID, secret_key=_AWS_SECRET, when=_AWS_WHEN)

    def signed_of(h: Dict[str, str]) -> str:
        auth = h["Authorization"]
        return auth.split("SignedHeaders=")[1].split(",")[0]

    got = "%s || %s" % (signed_of(with_both), signed_of(minimal))
    expect = ("host;x-amz-content-sha256;x-amz-date;x-amz-security-token"
              " || host;x-amz-date")
    return got == expect, got, expect


CASES = [
    ("AWS 官方 ListUsers 向量", _case_listusers),
    ("空载荷 sha256 常量", _case_empty_payload_hash),
    ("查询串编码后排序", _case_query_sorting),
    ("TikTok 代理 URL 规范化", _case_tiktok_url_shape),
    ("签名密钥派生链", _case_signing_key_vector),
    ("显式 content_sha256 优先", _case_content_sha256_override),
    ("x-amz-* 头必须进签名", _case_amz_headers_must_be_signed),
]


def selftest() -> int:
    print("=" * 78)
    print("SigV4 自证（AWS 官方向量 + 本地不变式）")
    print("=" * 78)
    passed = 0
    for name, fn in CASES:
        try:
            ok, got, expect = fn()
        except Exception as exc:  # noqa: BLE001
            print("  [ERR] %-26s 抛异常: %s" % (name, exc))
            continue
        print("  [%s] %-26s" % ("OK " if ok else "!! ", name))
        if not ok:
            print("        实际: %s" % got)
            print("        期望: %s" % expect)
        passed += bool(ok)
    print("-" * 78)
    print("  %d/%d 通过" % (passed, len(CASES)))
    return 0 if passed == len(CASES) else 1


if __name__ == "__main__":
    sys.exit(selftest())

File: lib/sigv4.js
'use strict';
/**
 * SigV4 签名器（火山引擎兼容）—— **零第三方依赖**，只用 node:crypto。
 *
 * 为什么手写而不是装 aws-sdk
 * --------------------------
 * 火山引擎 ImageX 的 OpenAPI 用的是 AWS SigV4 **兼容格式**，签名 scope 是
 * `region=i18n` / `service=imagex`（口径来自真实抓包的 Authorization 头）。
 * 目标执行节点是一台干净的托管小机器，不值得为一个确定性算法拖进
 * aws-sdk 那一坨依赖。
 *
 * 正确性如何保证
 * --------------
 * 算法是确定性的，所以可以用 **AWS 官方文档的已知测试向量**离线自证，
 * 不需要任何真实凭据、不产生任何网络请求：
 *
 *     node lib/sigv4.js
 *
 * 期望 `7/7 通过`。这条自证在拿到新鲜 cookie 之前就能跑，把「签名对不对」
 * 从「上服务器碰运气」变成「本机秒验」。
 *
 * ⚠️ 本文件是 `reference/sigv4.py` 的逐行移植。改动任意一侧都必须让另一侧
 *    的自证仍然全绿 —— 两边的用例是同一组，就是为此准备的。
 */
const crypto = require('node:crypto');

const ALGORITHM = 'AWS4-HMAC-SHA256';
const TERMINATOR = 'aws4_request';

/** RFC 3986 unreserved：A-Z a-z 0-9 - _ . ~ 之外一律百分号编码 */
const UNRESERVED = '-_.~';

/** 空载荷的 sha256 —— GET 也用这个，写错成 "" 是最常见的坑 */
const EMPTY_SHA256 = crypto.createHash('sha256').update('').digest('hex');

/** RFC 3986 编码，十六进制大写（SigV4 要求大写）。 */
function uriEncode(value, encodeSlash = true) {
  let out = encodeURIComponent(String(value))
    .replace(/[!'()*]/g, (c) => '%' + c.charCodeAt(0).toString(16).toUpperCase());
  if (!encodeSlash) out = out.replace(/%2F/g, '/');
  return out;
}

function sha256Hex(data) {
  const buf = Buffer.isBuffer(data) ? data : Buffer.from(String(data), 'utf8');
  return crypto.createHash('sha256').update(buf).digest('hex');
}

/**
 * 手写 URL 拆分，**刻意不用 new URL()** —— 后者会重排/重新编码查询串，
 * 而签名要求「先编码、再按编码后的字节序排序」，归一化必须由我们自己控制。
 */
function splitUrl(url) {
  const s = String(url);
  const m = /^([a-zA-Z][a-zA-Z0-9+.-]*):\/\/([^/?#]*)([^?#]*)(?:\?([^#]*))?/.exec(s);
  if (!m) throw new Error('无法解析 URL: ' + s);
  return { scheme: m[1], netloc: m[2], path: m[3] || '', query: m[4] || '' };
}

const decodeQueryPart = (s) => decodeURIComponent(String(s).replace(/\+/g, ' '));

/** 解析查询串，保留空值参数（等价 Python 的 parse_qsl(keep_blank_values=True)）。 */
function parseQuery(query) {
  if (!query) return [];
  return query.split('&').filter((p) => p !== '').map((pair) => {
    const i = pair.indexOf('=');
    return i === -1
      ? [decodeQueryPart(pair), '']
      : [decodeQueryPart(pair.slice(0, i)), decodeQueryPart(pair.slice(i + 1))];
  });
}

/**
 * 查询串归一化：先编码，再按**编码后的键**做字节序升序。
 *
 * ⚠️ 排序必须在编码之后进行 —— 大写字母（`A`=65）排在小写（`s`=115）之前，
 * 所以 `Action, FileSize, ServiceId, Version, device_platform, s`
 * 才是正确顺序。按原始串排序会得到不同结果、签名必然错。
 */
function canonicalQueryString(url) {
  const pairs = parseQuery(splitUrl(url).query);
  if (!pairs.length) return '';
  const encoded = pairs.map(([k, v]) => [uriEncode(k), uriEncode(v)]);
  encoded.sort(comparePairs);
  return encoded.map(([k, v]) => `${k}=${v}`).join('&');
}

function comparePairs(a, b) {
  if (a[0] !== b[0]) return a[0] < b[0] ? -1 : 1;
  if (a[1] !== b[1]) return a[1] < b[1] ? -1 : 1;
  return 0;
}

/** 键小写、值折叠连续空白并去首尾（SigV4 对值的规范化要求）。 */
function normalizeHeaders(headers) {
  const out = {};
  for (const [k, v] of Object.entries(headers || {})) {
    out[String(k).toLowerCase().trim()] = String(v).split(/\s+/).filter(Boolean).join(' ');
  }
  return out;
}

const VOLATILE_HEADERS = ['authorization', 'x-amz-date', 'x-amz-security-token', 'x-amz-content-sha256'];

/**
 * 返回 { canonical, signedHeaders }。
 *
 * 拼接后的形状（注意查询串为空时那一行是空的，且头块末尾会多出一个空行）：
 *
 *     GET
 *     /
 *
 *     content-type:...
 *     host:iam.amazonaws.com
 *     x-amz-date:20150830T123600Z
 *
 *     content-type;host;x-amz-date
 *     e3b0c442...
 */
function buildCanonicalRequest(method, url, headers, payloadSha) {
  const parsed = splitUrl(url);
  const body = Object.assign({}, headers);
  if (!Object.keys(body).some((k) => k.toLowerCase() === 'host')) body.host = parsed.netloc;

  const lower = normalizeHeaders(body);
  const keys = Object.keys(lower).sort();
  const signedHeaders = keys.join(';');
  const canonHeaders = keys.map((k) => `${k}:${lower[k]}\n`).join('');

  const canonical = [
    String(method).toUpperCase(),
    uriEncode(parsed.path || '/', false),
    canonicalQueryString(url),
    canonHeaders,
    signedHeaders,
    payloadSha,
  ].join('\n');

  return { canonical, signedHeaders };
}

/** AWS4 + secret → HMAC 链 → 派生签名密钥（4 轮）。 */
function deriveSigningKey(secretKey, dateStamp, region, service) {
  let key = Buffer.from('AWS4' + secretKey, 'utf8');
  for (const msg of [dateStamp, region, service, TERMINATOR]) {
    key = crypto.createHmac('sha256', key).update(Buffer.from(msg, 'utf8')).digest();
  }
  return key;
}

const pad2 = (n) => String(n).padStart(2, '0');

/** ISO8601 的 `YYYYMMDDTHHMMSSZ`（必须是 UTC）。 */
function amzDateOf(date) {
  return `${date.getUTCFullYear()}${pad2(date.getUTCMonth() + 1)}${pad2(date.getUTCDate())}` +
    `T${pad2(date.getUTCHours())}${pad2(date.getUTCMinutes())}${pad2(date.getUTCSeconds())}Z`;
}

/**
 * 对一次请求签名，返回**需要补进请求头**的字段（标准 HTTP 大小写形式）。
 *
 * `when` 必须是 UTC 的 Date。默认取当前时间；但对接 TikTok 时应当显式传入
 * STS 的 `CurrentTime` —— 节点机器时钟漂一点就会让签名落在 STS 有效窗口外，
 * 而报错只会是含糊的签名失败。
 *
 * `signContentSha256` —— 是否附带并签名 `X-Amz-Content-Sha256`。
 * ⚠️ AWS 规范规定**所有 `x-amz-*` 头都必须进签名**，所以「发不发」和
 * 「签不签」是同一个开关，不存在「发了但不签」这个中间态。
 *   - false（默认）：完全不发这个头。SigV4 的最小正确形态，AWS 官方示例的形态。
 *   - true：附带并签名。TikTok 的真实抓包里这个头**是存在的**，
 *     所以对接 TikTok 的 CommitImageUpload 要传 true。
 * `sessionToken` 一旦提供就必然附带并签名 —— STS 凭据缺它必被拒。
 */
function sign(opts) {
  const {
    method,
    url,
    accessKey,
    secretKey,
    headers = {},
    payload = Buffer.alloc(0),
    sessionToken = null,
    region = 'i18n',
    service = 'imagex',
    when = null,
    contentSha256 = null,
    signContentSha256 = false,
  } = opts || {};

  const date = when ? new Date(when) : new Date();
  const amzDate = amzDateOf(date);
  const dateStamp = amzDate.slice(0, 8);

  const payloadBuf = Buffer.isBuffer(payload) ? payload : Buffer.from(String(payload), 'utf8');
  const payloadSha = contentSha256 || sha256Hex(payloadBuf);

  const hdrs = {};
  for (const [k, v] of Object.entries(headers || {})) hdrs[String(k)] = String(v);
  for (const k of Object.keys(hdrs)) {
    if (VOLATILE_HEADERS.includes(k.toLowerCase())) delete hdrs[k];
  }
  if (!Object.keys(hdrs).some((k) => k.toLowerCase() === 'host')) hdrs.host = splitUrl(url).netloc;
  hdrs['X-Amz-Date'] = amzDate;
  if (signContentSha256) hdrs['X-Amz-Content-Sha256'] = payloadSha;
  if (sessionToken) hdrs['X-Amz-Security-Token'] = sessionToken;

  const { canonical, signedHeaders } = buildCanonicalRequest(method, url, hdrs, payloadSha);

  const scope = `${dateStamp}/${region}/${service}/${TERMINATOR}`;
  const stringToSign = [ALGORITHM, amzDate, scope, sha256Hex(canonical)].join('\n');

  const signingKey = deriveSigningKey(secretKey, dateStamp, region, service);
  const signature = crypto
    .createHmac('sha256', signingKey)
    .update(Buffer.from(stringToSign, 'utf8'))
    .digest('hex');

  const out = {
    Authorization: `${ALGORITHM} Credential=${accessKey}/${scope}, ` +
      `SignedHeaders=${signedHeaders}, Signature=${signature}`,
    'X-Amz-Date': amzDate,
  };
  if (signContentSha256) out['X-Amz-Content-Sha256'] = payloadSha;
  if (sessionToken) out['X-Amz-Security-Token'] = sessionToken;
  return out;
}

// ---------------------------------------------------------------------------
// 自证：AWS 官方文档的已知测试向量
//
// 来源：AWS 文档 "Examples of the complete Version 4 signing process" 里的
// ListUsers 示例。凭据与签名都是公开的示例值，不是真实密钥。
//
// 之所以能拿它当脚手架：火山引擎是 SigV4 *兼容* 实现，签名算法本身完全相同，
// 只有 scope 里的 region/service 两个字符串不同。所以只要这组向量过了，
// 就能确定 canonical request / 密钥派生 / 签名三处都没有写错。
// ---------------------------------------------------------------------------
const AWS_KEY_ID = 'AKIDEXAMPLE';
const AWS_SECRET = 'wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY';
const AWS_WHEN = new Date(Date.UTC(2015, 7, 30, 12, 36, 0));

function caseListUsers() {
  const out = sign({
    method: 'GET',
    url: 'https://iam.amazonaws.com/?Action=ListUsers&Version=2010-05-08',
    headers: {
      'content-type': 'application/x-www-form-urlencoded; charset=utf-8',
      host: 'iam.amazonaws.com',
    },
    payload: Buffer.alloc(0),
    accessKey: AWS_KEY_ID,
    secretKey: AWS_SECRET,
    region: 'us-east-1',
    service: 'iam',
    when: AWS_WHEN,
  });
  const expect = 'AWS4-HMAC-SHA256 Credential=AKIDEXAMPLE/20150830/us-east-1/iam/aws4_request, ' +
    'SignedHeaders=content-type;host;x-amz-date, ' +
    'Signature=5d672d79c15b13162d9279b0855cfba6789a8edb4c82c400e06b5924a6f2b5d7';
  return [out.Authorization === expect, out.Authorization, expect];
}

function caseEmptyPayloadHash() {
  const expect = 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855';
  return [EMPTY_SHA256 === expect, EMPTY_SHA256, expect];
}

function caseQuerySorting() {
  const url = 'https://x/?s=zzz&device_platform=web&Version=2018-08-01' +
    '&ServiceId=n2703mo9gi&FileSize=45431&Action=ApplyImageUpload';
  const got = canonicalQueryString(url);
  const expect = 'Action=ApplyImageUpload&FileSize=45431&ServiceId=n2703mo9gi' +
    '&Version=2018-08-01&device_platform=web&s=zzz';
  return [got === expect, got, expect];
}

function caseTiktokUrlShape() {
  const url = 'https://ads.tiktok.com/creative/creativestudio/upload-proxy' +
    '?Action=ApplyImageUpload&Version=2018-08-01' +
    '&ServiceId=n2703mo9gi&FileSize=45431&s=m0kza0mqao&device_platform=web';
  const got = canonicalQueryString(url);
  const expect = 'Action=ApplyImageUpload&FileSize=45431&ServiceId=n2703mo9gi' +
    '&Version=2018-08-01&device_platform=web&s=m0kza0mqao';
  return [got === expect, got, expect];
}

function caseSigningKeyVector() {
  const key = deriveSigningKey(
    'wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY', '20120215', 'us-east-1', 'iam');
  const expect = 'f4780e2d9f65fa895f9c67b32ce1baf0b0d8a43505a000a1a9e090d414db404d';
  return [key.toString('hex') === expect, key.toString('hex'), expect];
}

function caseContentSha256Override() {
  const url = 'https://ads.tiktok.com/creative/creativestudio/upload-proxy?Action=CommitImageUpload';
  const payload = Buffer.from('{"SessionKey":"abc"}', 'utf8');
  const forced = sha256Hex(payload);
  const out = sign({
    method: 'POST', url, payload, contentSha256: forced, signContentSha256: true,
    accessKey: AWS_KEY_ID, secretKey: AWS_SECRET, when: AWS_WHEN,
  });
  const ok = out['X-Amz-Content-Sha256'] === forced && forced !== EMPTY_SHA256;
  return [ok, out['X-Amz-Content-Sha256'], forced];
}

function caseAmzHeadersMustBeSigned() {
  const url = 'https://ads.tiktok.com/creative/creativestudio/upload-proxy?Action=ApplyImageUpload';
  const signedOf = (h) => h.Authorization.split('SignedHeaders=')[1].split(',')[0];
  const both = sign({
    method: 'GET', url, payload: Buffer.alloc(0), signContentSha256: true,
    sessionToken: 'STS2example', accessKey: AWS_KEY_ID, secretKey: AWS_SECRET, when: AWS_WHEN,
  });
  const minimal = sign({
    method: 'GET', url, payload: Buffer.alloc(0),
    accessKey: AWS_KEY_ID, secretKey: AWS_SECRET, when: AWS_WHEN,
  });
  const got = `${signedOf(both)} || ${signedOf(minimal)}`;
  const expect = 'host;x-amz-content-sha256;x-amz-date;x-amz-security-token || host;x-amz-date';
  return [got === expect, got, expect];
}

const CASES = [
  ['AWS 官方 ListUsers 向量', caseListUsers],
  ['空载荷 sha256 常量', caseEmptyPayloadHash],
  ['查询串编码后排序', caseQuerySorting],
  ['TikTok 代理 URL 规范化', caseTiktokUrlShape],
  ['签名密钥派生链', caseSigningKeyVector],
  ['显式 content_sha256 优先', caseContentSha256Override],
  ['x-amz-* 头必须进签名', caseAmzHeadersMustBeSigned],
];

function selftest() {
  console.log('='.repeat(78));
  console.log('SigV4 自证（AWS 官方向量 + 本地不变式）');
  console.log('='.repeat(78));
  let passed = 0;
  for (const [name, fn] of CASES) {
    let ok = false, got = '', expect = '';
    try {
      [ok, got, expect] = fn();
    } catch (err) {
      console.log(`  [ERR] ${name.padEnd(26)} 抛异常: ${err.message}`);
      continue;
    }
    console.log(`  [${ok ? 'OK ' : '!! '}] ${name}`);
    if (!ok) {
      console.log(`        实际: ${got}`);
      console.log(`        期望: ${expect}`);
    }
    if (ok) passed += 1;
  }
  console.log('-'.repeat(78));
  console.log(`  ${passed}/${CASES.length} 通过`);
  return passed === CASES.length ? 0 : 1;
}

module.exports = {
  ALGORITHM,
  TERMINATOR,
  EMPTY_SHA256,
  uriEncode,
  sha256Hex,
  splitUrl,
  canonicalQueryString,
  buildCanonicalRequest,
  deriveSigningKey,
  sign,
  selftest,
};

if (require.main === module) process.exit(selftest());
